patches/tModLoader/Terraria/ModLoader/UI/ModBrowser/ModDownloadItem.cs

```diff
@@ -4,93 +4,110 @@
+	public bool IsInstalled => Installed != null;
+	public bool IsEnabled => IsInstalled && Installed.Enabled;
+
+	public ModDownloadItem(string displayName, string name, string version, string author, string modReferences, ModSide modSide, string modIconUrl, string publishId, int downloads, int hot, DateTime timeStamp, string modloaderversion, string homepage, string ownerId, string[] referencesById)
 	{
 		ModName = name;
 		DisplayName = displayName;
 		DisplayNameClean = string.Join("", ChatManager.ParseMessage(displayName, Color.White).Where(x => x.GetType() == typeof(TextSnippet)).Select(x => x.Text));
-		PublishId = publishId;
+		PublishId = new ModPubId_t { m_ModPubId = publishId };
+		OwnerId = ownerId;
 
 		Author = author;
-		ModReferences = modReferences;
+		ModReferencesBySlug = modReferences;
+		ModReferenceByModId = Array.ConvertAll(referencesById, x => new ModPubId_t() { m_ModPubId = x});
 		ModSide = modSide;
 		ModIconUrl = modIconUrl;
 		Downloads = downloads;
 		Hot = hot;
 		Homepage = homepage;
 		TimeStamp = timeStamp;
-		HasUpdate = hasUpdate;
-		UpdateIsDowngrade = updateIsDowngrade;
-		NeedsGameRestart = needsRestart;
-		Installed = installed;
 		Version = version;
 		ModloaderVersion = modloaderversion;
 	}
 
-	internal ModDownloadItem(string displayName, string publishId, LocalMod installed)
+	internal void UpdateInstallState()
 	{
-		DisplayName = displayName;
-		DisplayNameClean = string.Join("", ChatManager.ParseMessage(displayName, Color.White).Where(x => x.GetType() == typeof(TextSnippet)).Select(x => x.Text));
-		PublishId = publishId;
-		Installed = installed;
+		// Remember this method is blocking, it does network stuff... - DarioDaf
+
+		// Check against installed mods for updates.
+		//TODO: This should assess the source of the ModDownloadItem and ensure matches with the active SocialBrowserModule instance for safety, but eh.
+		Installed = Interface.modBrowser.SocialBackend.IsItemInstalled(this.ModName);
+
+		NeedUpdate = Installed != null && Interface.modBrowser.SocialBackend.DoesItemNeedUpdate(this.PublishId, Installed, new System.Version(this.Version));
+		// The below line is to identify the transient state where it isn't installed, but Steam considers it as such - Solxan
+		// Steam keeps a cache once a download starts, and doesn't clean up cache until game close, which gets very confusing.
+		AppNeedRestartToReinstall = Installed == null && Interface.modBrowser.SocialBackend.DoesAppNeedRestartToReinstallItem(this.PublishId);
 	}
 
-	internal Task InnerDownloadWithDeps()
+	public override bool Equals(object obj) => this.Equals(obj as ModDownloadItem);
+
+	// Custom Equality for Mod Browser efficiency
+	public bool Equals(ModDownloadItem item)
```

Comment:
Will add `GetHashCode` and work off a tuple to keep in sync with Equals, not sure where it's used tho @Solxanich was this added and then made deprecated?

Reply:
Added comment, needed for unexplained behavioural issues with List<ModDownloadItem>.Contains(ModDownloadItem).
Can be removed if fixed through testing.

Not sure what overriding GetHashCode does for us outside of optimization to ignore information, but doesn't hurt I suppose.

Reply:
Hash not matching but equal returning true (which is the case since hash by default I think is the location of the object hashed) can result in random stuff when working with containers (List should not, but it's technically runtime dependent, so overriding only one is technically undefined behaviour)